tools/swift-inspect/Sources/SwiftInspectHeapWalk/SwiftInspectHeapWalk.cpp

```diff
@@ -0,0 +1,184 @@
+#if defined(_WIN32)
+
+#pragma comment(lib, "swiftCore.lib")
+
+#include <windows.h>
+#include <stdio.h>
+
+namespace {
+struct HeapEntry {
+  uintptr_t address;
+  uintptr_t size;
+};
+} // anonymous namespace
+
+#define BUF_SIZE 512
+#define BUF_NUM_ENTRIES (BUF_SIZE / sizeof(HeapEntry))
+#define SHARED_MEM_NAME "Local\\SwiftInspectFileMapping"
+#define READ_EVENT_NAME "Local\\SwiftInspectReadEvent"
+#define WRITE_EVENT_NAME "Local\\SwiftInspectWriteEvent"
+#define WAIT_TIMEOUT_MS 30000
+
+static int heapWalk() {
+  HANDLE hMapFile = OpenFileMappingA(FILE_MAP_ALL_ACCESS, false,
+                                     SHARED_MEM_NAME);
+  if (hMapFile == NULL) {
+    printf("OpenFileMapping failed\n");
+    return 1;
+  }
+  HeapEntry *buf = (HeapEntry *)
+    MapViewOfFile(hMapFile,
+                  FILE_MAP_ALL_ACCESS,
+                  0,
+                  0,
+                  BUF_SIZE);
+  if (buf == NULL) {
+    printf("MapViewOfFile failed\n");
+    CloseHandle(hMapFile);
+    return 1;
+  }
+  memset(buf, 0, BUF_SIZE);
+  HANDLE writeEvent = OpenEventA(EVENT_ALL_ACCESS,
+                                 false,
+                                 WRITE_EVENT_NAME);
+  if (writeEvent == NULL) {
+    printf("OpenEventA failed\n");
+    CloseHandle(hMapFile);
+    UnmapViewOfFile(buf);
+    return 1;
+  }
+  HANDLE readEvent = OpenEventA(EVENT_ALL_ACCESS,
+                                false,
+                                READ_EVENT_NAME);
+  if (readEvent == NULL) {
+    printf("OpenEventA failed\n");
+    CloseHandle(hMapFile);
+    UnmapViewOfFile(buf);
+    CloseHandle(writeEvent);
+    return 1;
+  }
+
+  auto cleanupIPC = [&] {
+    CloseHandle(hMapFile);
+    UnmapViewOfFile(buf);
+    CloseHandle(writeEvent);
+    CloseHandle(readEvent);
```

Comment:
Use of `scoped_handle` should take care of three of these and leave only `MapViewOfFile`.  I wonder if we can use `std::unique_ptr` with a custom deleter for that.